accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -616,6 +609,442 @@ It is a warning to use a null aware operator (`?.`, `?[]`, `?..`, `??`, `??=`, o
 It is a warning to use the null check operator (`!`) on an expression of type
 `T` if `T` is **strictly non-nullable** .
 
+### Local variables and definite (un)assignment.
+
+As part of the null safety release, errors for local variables are specified to
+take into account **definite assignment** and **definite unassignment** (see the
+section on Definite Assignment below).  We say that a variable is **potentially
+assigned** if it is not **definitely unassigned**, and that a variable is
+**potentially unassigned** if it is not **definitely assigned**.
+
+In all cases in this section, errors that are described as occurring on reads of
+a variable are intended to apply to all form of reads, including indirectly as
+part of composite assignment operators, as well as via pre and post-fix
+operators.  Similarly, errors that are described as occurring on writes of a
+variable are intended to apply to all form of writes.
+
+It is a compile time error to assign a value to a `final`, non-`late` local
+variable which is **potentially assigned**.  Thus, it is *not* an error to
+assign to a **definitely unassigned** `final` local variable.
+
+It is a compile time error to assign a value to a `final`, `late` local variable
+if it is **definitely assigned**. Thus, it is *not* an error to assign to a
+**potentially unassigned** `final`, `late` local variable.
+
+*Note that a variable is always considered **definitely assigned** and not
+**definitely unassigned** if it has an explicit initializer, or an implicit
+initializer as part of a larger construct (e.g. the loop variable in a `for in`
+construct).*
```

Comment:
This is yet another place where model 3 (https://github.com/dart-lang/language/issues/946#issuecomment-649229421) would require an exception: With `var x = null;`, `x` would be considered definitely unassigned, or no assignment after the declaration could be considered an initializing assignment.

Reply:
Yes, I'm thinking we'll need to have intiailziation/typing separate from assignment.

A variable is either typed, untyped (with nullable hint or not), or initialization-typed (or maybe also invalid-typed if we want to continue after incompatible joins).

If it's untyped and assigned to, then it becomes initialization-typed. We can join initialization-typed paths for variables, maybe untyped ones too (if we want to).